pkg/chunk/aws/s3_storage_client.go

```diff
@@ -53,43 +74,133 @@ func (cfg *S3Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<bucket-name> to use a mock in-memory implementation.")
 	f.BoolVar(&cfg.S3ForcePathStyle, prefix+"s3.force-path-style", false, "Set this to `true` to force the request to use path-style addressing.")
 	f.StringVar(&cfg.BucketNames, prefix+"s3.buckets", "", "Comma separated list of bucket names to evenly distribute chunks over. Overrides any buckets specified in s3.url flag")
+
+	f.StringVar(&cfg.Endpoint, prefix+"s3.endpoint", "", "S3 Endpoint to connect to.")
+	f.StringVar(&cfg.Region, prefix+"s3.region", "", "AWS region to use.")
+	f.StringVar(&cfg.AccessKeyID, prefix+"s3.access-key-id", "", "AWS Access Key ID")
+	f.StringVar(&cfg.SecretAccessKey, prefix+"s3.secret-access-key", "", "AWS Secret Access Key")
+	f.BoolVar(&cfg.SSEEncryption, prefix+"s3.sse-encryption", false, "Enable AES256 AWS Server Side Encryption")
+
+	f.DurationVar(&cfg.HTTPConfig.IdleConnTimeout, prefix+"s3.http.idle-conn-timeout", 90*time.Second, "The maximum amount of time an idle connection will be held open.")
+	f.DurationVar(&cfg.HTTPConfig.ResponseHeaderTimeout, prefix+"s3.http.response-header-timeout", 0, "If non-zero, specifies the amount of time to wait for a server's response headers after fully writing the request.")
+	f.BoolVar(&cfg.HTTPConfig.InsecureSkipVerify, prefix+"s3.http.insecure-skip-verify", false, "Set to false to skip verifying the certificate chain and hostname.")
 }
 
 type S3ObjectClient struct {
-	bucketNames []string
-	S3          s3iface.S3API
-	delimiter   string
+	bucketNames     []string
+	S3              s3iface.S3API
+	delimiter       string
+	sseEncryption   *string
+	putUserMetadata map[string]*string
 }
 
 // NewS3ObjectClient makes a new S3-backed ObjectClient.
 func NewS3ObjectClient(cfg S3Config, delimiter string) (*S3ObjectClient, error) {
-	if cfg.S3.URL == nil {
-		return nil, fmt.Errorf("no URL specified for S3")
+	s3Config, bucketNames, err := buildS3Config(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to build s3 config: %v", err)
```

Comment:
Golang best practice is to never capitalise error messages. Could you change this and the error below, please?